File: app/src/main/java/net/squanchy/notification/NotificationCreator.java
package net.squanchy.notification;

import android.app.Notification;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.content.pm.ApplicationInfo;
import android.content.res.Resources;
import android.graphics.BitmapFactory;
import android.graphics.Color;
import android.support.v4.app.NotificationCompat;
import android.support.v4.app.TaskStackBuilder;
import android.support.v4.content.ContextCompat;
import android.text.TextUtils;

import java.util.ArrayList;
import java.util.List;

import net.squanchy.R;
import net.squanchy.eventdetails.EventDetailsActivity;
import net.squanchy.home.HomeActivity;
import net.squanchy.schedule.domain.view.Event;
import net.squanchy.schedule.domain.view.Place;
import net.squanchy.speaker.domain.view.Speaker;

import static net.squanchy.support.lang.Lists.map;

public class NotificationCreator {

    private static final String GROUP_KEY_NOTIFY_SESSION = "group_key_notify_session";

    // pulsate every 1 second, indicating a relatively high degree of urgency
    private static final int NOTIFICATION_LED_ON_MS = 100;
    private static final int NOTIFICATION_LED_OFF_MS = 1000;
    private static final String EMPTY_PLACE_NAME = "";
    private static final String ARGB_TRANSPARENT = "#00000000";

    private final Context context;

    public NotificationCreator(Context context) {
        this.context = context;
    }

    public List<Notification> createFrom(List<Event> events) {
        List<Notification> notifications = new ArrayList<>();

        for (Event event : events) {
            notifications.add(createFrom(event));
        }

        if (events.size() > 1) {
            notifications.add(createSummaryNotification(events));
        }
        return notifications;
    }

    private Notification createFrom(Event event) {
        NotificationCompat.Builder notificationBuilder = createDefaultBuilder(1);
        notificationBuilder
                .setContentIntent(createPendingIntentForSingleEvent(event.getId()))
                .setContentTitle(event.getTitle())
                .setColor(getTrackColor(event))
                .setWhen(event.getStartTime().toDateTime().getMillis())
                .setShowWhen(true)
                .setGroup(GROUP_KEY_NOTIFY_SESSION);

        String placeName = getPlaceName(event);
        if (!placeName.isEmpty()) {
            notificationBuilder.setContentText(placeName);
        }

        NotificationCompat.BigTextStyle richNotification = createBigTextRichNotification(notificationBuilder, event);

        return richNotification.build();
    }

    public Notification createFromProximity(String proximityId, String bigText, String smallText) {
        Resources resources = context.getResources();

        NotificationCompat.WearableExtender extender = new NotificationCompat.WearableExtender();
        extender.setBackground(BitmapFactory.decodeResource(resources, R.drawable.notification_background));

        NotificationCompat.Builder builder =
                new NotificationCompat.Builder(context)
                        .setContentIntent(createPendingIntentForProximityNotification(proximityId))
                        .setContentTitle(getApplicationName(context))
                        .setContentText(smallText)
                        .setSmallIcon(R.drawable.ic_place_white_24dp)
                        .setLights(
                                ContextCompat.getColor(context, R.color.notification_led_color),
                                NOTIFICATION_LED_ON_MS,
                                NOTIFICATION_LED_OFF_MS
                        )
                        .setLargeIcon(BitmapFactory.decodeResource(resources, R.mipmap.ic_launcher))
                        .setStyle(new NotificationCompat.BigTextStyle()
                                .bigText(bigText))
                        .setPriority(Notification.PRIORITY_MAX)
                        .setAutoCancel(true)
                        .extend(extender);
        return builder.build();
    }

    private String getApplicationName(Context context) {
        ApplicationInfo applicationInfo = context.getApplicationInfo();
        int stringId = applicationInfo.labelRes;
        return stringId == 0 ? applicationInfo.nonLocalizedLabel.toString() : context.getString(stringId);
    }

    private Notification createSummaryNotification(List<Event> events) {
        NotificationCompat.Builder summaryBuilder = createDefaultBuilder(events.size());
        summaryBuilder
                .setContentIntent(createPendingIntentForMultipleEvents())
                .setContentTitle(createSummaryTitle(events.size()))
                .setGroup(GROUP_KEY_NOTIFY_SESSION)
                .setGroupSummary(true)
                .setLocalOnly(true);

        NotificationCompat.InboxStyle richNotification = createInboxStyleRichNotification(summaryBuilder, events);

        return richNotification.build();
    }

    private PendingIntent createPendingIntentForProximityNotification(String proximityId) {
        TaskStackBuilder taskBuilder = createBaseTaskStackBuilder();
        Intent homeIntent = HomeActivity.createProximityIntent(context, proximityId);
        taskBuilder.addNextIntent(homeIntent);

        return taskBuilder.getPendingIntent(0, PendingIntent.FLAG_CANCEL_CURRENT);
    }

    private NotificationCompat.Builder createDefaultBuilder(int talksCount) {
        Resources resources = context.getResources();

        NotificationCompat.WearableExtender extender = new NotificationCompat.WearableExtender();
        extender.setBackground(BitmapFactory.decodeResource(resources, R.drawable.notification_background));

        return new NotificationCompat.Builder(context)
                .setTicker(
                        context.getResources().getQuantityString(
                                R.plurals.event_notification_ticker,
                                talksCount,
                                talksCount
                        )
                )
                .setDefaults(Notification.DEFAULT_SOUND | Notification.DEFAULT_VIBRATE)
                .setLights(
                        ContextCompat.getColor(context, R.color.notification_led_color),
                        NOTIFICATION_LED_ON_MS,
                        NOTIFICATION_LED_OFF_MS
                )
                .setSmallIcon(R.drawable.ic_stat_notification)
                .setPriority(Notification.PRIORITY_MAX)
                .setAutoCancel(true)
                .extend(extender);
    }

    private PendingIntent createPendingIntentForSingleEvent(String eventId) {
        TaskStackBuilder taskBuilder = createBaseTaskStackBuilder();
        Intent eventDetailIntent = EventDetailsActivity.createIntent(context, eventId);
        taskBuilder.addNextIntent(eventDetailIntent);

        return taskBuilder.getPendingIntent(0, PendingIntent.FLAG_CANCEL_CURRENT);
    }

    private String getPlaceName(Event event) {
        return event.getPlace().map(Place::getName).or(EMPTY_PLACE_NAME);
    }

    private int getTrackColor(Event event) {
        return event.getTrack()
                .map(track -> Color.parseColor(track.getAccentColor().or(ARGB_TRANSPARENT)))
                .or(Color.TRANSPARENT);
    }

    private PendingIntent createPendingIntentForMultipleEvents() {
        TaskStackBuilder taskBuilder = createBaseTaskStackBuilder();
        return taskBuilder.getPendingIntent(0, PendingIntent.FLAG_CANCEL_CURRENT);
    }

    private TaskStackBuilder createBaseTaskStackBuilder() {
        Intent homescreenIntent = new Intent(context, HomeActivity.class);
        homescreenIntent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
        return TaskStackBuilder.create(context)
                .addNextIntent(homescreenIntent);
    }

    private NotificationCompat.BigTextStyle createBigTextRichNotification(NotificationCompat.Builder notificationBuilder, Event event) {
        StringBuilder bigTextBuilder = new StringBuilder()
                .append(getSpeakerNamesFrom(event.getSpeakers()));

        String placeName = getPlaceName(event);
        if (!placeName.isEmpty()) {
            bigTextBuilder
                    .append('\n')
                    .append(context.getString(R.string.event_notification_starting_in, placeName));
        }

        return new NotificationCompat.BigTextStyle(notificationBuilder)
                .setBigContentTitle(event.getTitle())
                .bigText(bigTextBuilder.toString());
    }

    private String getSpeakerNamesFrom(List<Speaker> speakers) {
        String speakerNames = TextUtils.join(", ", map(speakers, Speaker::getName));
        return context.getString(R.string.event_notification_starting_by, speakerNames);
    }

    private NotificationCompat.InboxStyle createInboxStyleRichNotification(NotificationCompat.Builder notificationBuilder, List<Event> events) {
        String bigContentTitle = createSummaryTitle(events.size());
        NotificationCompat.InboxStyle richNotification = new NotificationCompat.InboxStyle(notificationBuilder)
                .setBigContentTitle(bigContentTitle);

        for (Event event : events) {
            if (event.getPlace().isPresent()) {
                richNotification.addLine(
                        context.getString(
                                R.string.room_event_notification,
                                event.getPlace().get().getName(),
                                event.getTitle()
                        )
                );
            } else {
                richNotification.addLine(event.getTitle());
            }
        }

        return richNotification;
    }

    private String createSummaryTitle(int talksCount) {
        return context.getString(R.string.event_notification_count_starting, talksCount);
    }
}

File: app/src/main/java/net/squanchy/notification/NotificationCreator.kt
package net.squanchy.notification

import android.annotation.TargetApi
import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.graphics.BitmapFactory
import android.graphics.Color
import android.os.Build
import androidx.core.app.NotificationCompat
import androidx.core.app.TaskStackBuilder
import androidx.core.content.ContextCompat
import net.squanchy.R
import net.squanchy.eventdetails.EventDetailsActivity
import net.squanchy.home.HomeActivity
import net.squanchy.schedule.domain.view.Event
import net.squanchy.speaker.domain.view.Speaker
import net.squanchy.support.android
import net.squanchy.support.lang.getOrThrow
import net.squanchy.support.lang.or

class NotificationCreator(private val context: Context) {

    fun createFrom(events: List<Event>): List<Notification> {
        if (android.isAtLeastOreo) {
            createChannel()
        }

        val notifications = events
            .map(::createFrom)
            .toMutableList()

        if (events.size > 1) {
            notifications.add(createSummaryNotification(events))
        }
        return notifications
    }

    @TargetApi(Build.VERSION_CODES.O)
    private fun createChannel() {
        val channel = NotificationChannel(
            EVENTS_ABOUT_TO_START_CHANNEL_ID,
            context.getString(R.string.event_notification_starting_channel_name),
            NotificationManager.IMPORTANCE_HIGH
        )
        channel.description = context.getString(R.string.event_notification_starting_channel_description)
        channel.enableLights(true)
        channel.lightColor = context.getColor(R.color.notification_led_color)

        val notificationManager = context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        notificationManager.createNotificationChannel(channel)
    }

    private fun createFrom(event: Event): Notification {
        val notificationBuilder = createDefaultBuilder(1)
        notificationBuilder
            .setContentIntent(createPendingIntentForSingleEvent(event.id))
            .setContentTitle(event.title)
            .setColor(getTrackColor(event))
            .setWhen(event.startTime.toDateTime().millis)
            .setShowWhen(true)
            .setGroup(GROUP_KEY_NOTIFY_SESSION)

        val placeName = getPlaceName(event)
        if (!placeName.isNullOrEmpty()) {
            notificationBuilder.setContentText(placeName)
        }

        val richNotification = createBigTextRichNotification(notificationBuilder, event)

        return richNotification.build()
    }

    private fun createSummaryNotification(events: List<Event>): Notification {
        val summaryBuilder = createDefaultBuilder(events.size)
        summaryBuilder
            .setContentIntent(createPendingIntentForMultipleEvents())
            .setContentTitle(createSummaryTitle(events.size))
            .setGroup(GROUP_KEY_NOTIFY_SESSION)
            .setGroupSummary(true)
            .setLocalOnly(true)

        val richNotification = createInboxStyleRichNotification(summaryBuilder, events)

        return richNotification.build()
    }

    private fun createDefaultBuilder(talksCount: Int): NotificationCompat.Builder {
        val resources = context.resources

        val extender = NotificationCompat.WearableExtender()
        extender.background = BitmapFactory.decodeResource(resources, R.drawable.notification_background)

        return NotificationCompat.Builder(context, EVENTS_ABOUT_TO_START_CHANNEL_ID)
            .setTicker(
                context.resources.getQuantityString(
                    R.plurals.event_notification_ticker,
                    talksCount,
                    talksCount
                )
            )
            .setDefaults(Notification.DEFAULT_SOUND or Notification.DEFAULT_VIBRATE)
            .setLights(
                ContextCompat.getColor(context, R.color.notification_led_color),
                NOTIFICATION_LED_ON_MS,
                NOTIFICATION_LED_OFF_MS
            )
            .setSmallIcon(R.drawable.ic_stat_notification)
            .setAutoCancel(true)
            .extend(extender)
    }

    private fun createPendingIntentForSingleEvent(eventId: String): PendingIntent {
        val taskBuilder = createBaseTaskStackBuilder()
        val eventDetailIntent = EventDetailsActivity.createIntent(context, eventId)
        taskBuilder.addNextIntent(eventDetailIntent)

        return taskBuilder.getPendingIntent(0, PendingIntent.FLAG_CANCEL_CURRENT)!!
    }

    private fun getPlaceName(event: Event): String? {
        return event.place.orNull()?.name
    }

    private fun getTrackColor(event: Event): Int {
        return event.track
            .map { Color.parseColor(it.accentColor ?: ARGB_TRANSPARENT) }
            .or(Color.TRANSPARENT)
    }

    private fun createPendingIntentForMultipleEvents(): PendingIntent {
        val taskBuilder = createBaseTaskStackBuilder()
        return taskBuilder.getPendingIntent(0, PendingIntent.FLAG_CANCEL_CURRENT)!!
    }

    private fun createBaseTaskStackBuilder(): TaskStackBuilder {
        val homescreenIntent = Intent(context, HomeActivity::class.java)
        homescreenIntent.flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK
        return TaskStackBuilder.create(context)
            .addNextIntent(homescreenIntent)
    }

    private fun createBigTextRichNotification(notificationBuilder: NotificationCompat.Builder, event: Event): NotificationCompat.BigTextStyle {
        val bigTextBuilder = StringBuilder()
            .append(getSpeakerNamesFrom(event.speakers))

        val placeName = getPlaceName(event)
        if (!placeName.isNullOrEmpty()) {
            bigTextBuilder
                .append('\n')
                .append(context.getString(R.string.event_notification_starting_in, placeName))
        }

        return NotificationCompat.BigTextStyle(notificationBuilder)
            .setBigContentTitle(event.title)
            .bigText(bigTextBuilder.toString())
    }

    private fun getSpeakerNamesFrom(speakers: List<Speaker>): String {
        val speakerNames = speakers.joinToString(separator = ", ", transform = { it.name })

        return context.getString(R.string.event_notification_starting_by, speakerNames)
    }

    private fun createInboxStyleRichNotification(
        notificationBuilder: NotificationCompat.Builder,
        events: List<Event>
    ): NotificationCompat.InboxStyle {
        val bigContentTitle = createSummaryTitle(events.size)
        val richNotification = NotificationCompat.InboxStyle(notificationBuilder)
            .setBigContentTitle(bigContentTitle)

        for (event in events) {
            if (event.place.isDefined()) {
                richNotification.addLine(
                    context.getString(
                        R.string.room_event_notification,
                        event.place.getOrThrow().name,
                        event.title
                    )
                )
            } else {
                richNotification.addLine(event.title)
            }
        }

        return richNotification
    }

    private fun createSummaryTitle(talksCount: Int): String {
        val quantityString = context.resources.getQuantityString(R.plurals.event_notification_count_starting, talksCount)
        return String.format(quantityString, talksCount)
    }

    companion object {
        private const val GROUP_KEY_NOTIFY_SESSION = "group_key_notify_session"
        private const val EVENTS_ABOUT_TO_START_CHANNEL_ID = "events_about_to_start"

        // pulsate every 1 second, indicating a relatively high degree of urgency
        private const val NOTIFICATION_LED_ON_MS = 100
        private const val NOTIFICATION_LED_OFF_MS = 1000
        private const val ARGB_TRANSPARENT = "#00000000"
    }
}
